src/frontend/src/utils/iiConnection.ts

```diff
@@ -229,20 +233,68 @@ export class Connection {
 
     const actor = await this.createActor(delegationIdentity);
 
+    const connection = new AuthenticatedConnection(
+      this.canisterId,
+      identity,
+      delegationIdentity,
+      userNumber,
+      actor
+    );
+
+    const attachmentInfo = identity.getAuthenticatorAttachment();
+    if (attachmentInfo !== undefined) {
+      this.updateKeyTypeIfNecessary(devices, attachmentInfo, connection);
+    }
+
     return {
       kind: "loginSuccess",
       userNumber,
-      connection: new AuthenticatedConnection(
-        this.canisterId,
-        // eslint-disable-next-line
-        identity,
-        delegationIdentity,
-        userNumber,
-        actor
-      ),
+      connection,
     };
   };
 
+  private updateKeyTypeIfNecessary(
+    devices: DeviceData[],
+    attachmentInfo: {
+      credentialId: ArrayBuffer;
+      authenticatorAttachment: AuthenticatorAttachment;
+    },
+    authenticatedConnection: AuthenticatedConnection
+  ) {
+    const device = devices.find((device) => {
+      const credentialId = device.credential_id[0];
+      if (credentialId === undefined) {
+        return false;
+      }
+      return bufferEqual(
+        Buffer.from(credentialId),
+        attachmentInfo.credentialId
+      );
+    });
+
+    // only update devices with key-type unknown
+    if (device !== undefined && hasOwnProperty(device.key_type, "unknown")) {
+      switch (attachmentInfo.authenticatorAttachment) {
+        case "cross-platform":
+          device.key_type = { cross_platform: null };
+          break;
+        case "platform":
+          device.key_type = { platform: null };
+          break;
+        default:
+          unreachable(attachmentInfo.authenticatorAttachment);
+          break;
+      }
+      // we purposely do not await the promise as we just optimistically update
+      // if it fails, no harm done
+      authenticatedConnection.update(device).catch((error) => {
```

Comment:
we're hoping that this'll only happen once, since after this succeeds `device` will have a `key_type` set and this will be skipped, right?

Reply:
Yes, exactly.